### ISSUE TITLE
VAE encode/decode messing up faces

### ISSUE THREAD

**Issue by aaron9412** [comfyanonymous/ComfyUI#673]
Hi, comfyui is awesome!!

I'm having a problem where any time the VAE recognizes a face, it gets distorted.
I can build a simple workflow (loadvae, vaedecode, vaeencode, previewimage) with an input image.
The output image will have a distorted face with no other distortions.
I'm not sure what to do about this. I have tried mse, ema, and anything, and they all produce the result.
Automatic1111 does not do this in img2img or inpainting, so I assume its something going on in comfy.
Windows10, latest comfy (downloaded 16MAY2023, idk where to find version number in comfy), impact and was nodes installed.
Any suggestions appreciated. Maybe there is a way to use a null VAE?

Have a great day!
-Aaron

**Comment by ltdrdata** [comment]
If you attach your workflow. It might be helpful.

**Comment by DivinoAG** [comment]
> Are you using models vae and not override? Most people override in A1111 cause model vaes are often broken. Try using custom vaes with models.

That seems to be the case for me, I tried many models that say they include embedded VAEs, and the results are almost universally worse than using the standard 1.5 VAE (`vae-ft-mse-840000-ema-pruned`). This might be worth checking.

**Comment by aaron9412** [comment]
I am using VAE loader with the common VAE models I think.
Please see my example below of comfy having an issue with a face but A111 does not -

Comfy Workflow:
![image](https://github.com/comfyanonymous/ComfyUI/assets/27147792/d8dbb4ec-5949-45e8-81f8-2a8fd5802789)
Input face:
![image](https://github.com/comfyanonymous/ComfyUI/assets/27147792/c5d6b414-cc96-488f-a40f-6013f5ba57e2)
Output face:
![image](https://github.com/comfyanonymous/ComfyUI/assets/27147792/855a76d4-37d1-4bef-b9df-1e09739c2421)

Same picture, same VAE in inpaint in A1111:
![image](https://github.com/comfyanonymous/ComfyUI/assets/27147792/6caa6527-97ba-494d-aa5a-02478383dabd)
Input face:
![image](https://github.com/comfyanonymous/ComfyUI/assets/27147792/75efb538-f51c-4b81-bca3-2bab95345c71)
Output face:
![image](https://github.com/comfyanonymous/ComfyUI/assets/27147792/e7aa2f61-5dbf-424e-932c-8acf14bfcde2)

The changes in the A111 face are from my hackjob inpaint I think, A1111 still has the same mouth, mustache, eyes, comfy has a messed up mustache and mouth with no other input except VAE encode/decode.

Hope this helps!

**Comment by aaron9412** [comment]
Here's the input image since it's a bit hard to replicate, it only happens to certain images.

![R](https://github.com/comfyanonymous/ComfyUI/assets/27147792/e5965c37-c53b-4e06-9bee-41b816514f68)

**Comment by WASasquatch** [comment]
This looks like it is related to issue I mentioned in the chat. The vae decoded raw images have a either applied which messes with all the feature lines of a image. This would cause issues in the decode process likely.

**Comment by ltdrdata** [comment]
I found that VAEEncodeForInpaint's weired action.
VAEEncode + SetLatentNoise is working well.

I thought the difference of VAEEncodeForInpaint and VAEEncode + SetLatentNoise is cropping for efficiency.
But it is not.

**Comment by pennyvc** [comment]
Latest update here, using WAS, Efficiency and Quality of life nodes. although this bug is not because of those custom nodes.
I get similar vae issues, vaces are almost always messed up but in addition. Vae encode and decode causes artifacts to appear. 
![vaebug2](https://github.com/comfyanonymous/ComfyUI/assets/128430314/bb28fa7e-2210-4650-9fd0-c81b4d24a07b)
Before and after vae encoding.

![vaebug1](https://github.com/comfyanonymous/ComfyUI/assets/128430314/dddfdb6a-21c7-419a-a031-d69067197368)
after 5 steps of sampler and additional vae encode.

Theres definately something going on with vae process, it tends to appear in darker images or areas with a high contrast sharp edge, Im assuming this is not normally spotted because of denoise process wiping it out but if you are just strait up working with image nodes and very low step sampling you end up with artifacts.

Can supply workflow if needed but custom nodes will bork.
![vaebug3](https://github.com/comfyanonymous/ComfyUI/assets/128430314/d6a6e19f-8a97-4a5a-a08e-39fb873de316)
heres an image of it. 
tested with overide Vae 'vae-ft-mse-840000-ema-pruned' and embedded vae

**Comment by pennyvc** [comment]
not something I think most would notice using like a straight img2img of bright images or just full noise on first sampler.
Yeh, judging from how latent scaling works its pretty lossy. Hopefully its as simple as cropping in 2 pixels or something but i doubt it.
also, lol, making pictures of the fungus forests from that movie. but i looove dark images which is how i keep stumbling into this vae issue :<

**Comment by WASasquatch** [comment]
> not something I think most would notice using like a straight img2img of bright images or just full noise on first sampler. Yeh, judging from how latent scaling works its pretty lossy. Hopefully its as simple as cropping in 2 pixels or something but i doubt it. also, lol, making pictures of the fungus forests from that movie. but i looove dark images which is how i keep stumbling into this vae issue :<

I feel like this explains the random borders that are hard to get rid of some times. That just creep up out of no where. it's not like you're all "Give me a picture frame". I think it just interprets the edges pixel issue and does a border. 

---

Also awesome! I tried doing the toxic jungle too! Though that was back in Disco Diffusion, and it was surprising how well it understood the prompt. Can tell it was picking up on resources around the film. I should try again in SD. Thanks for reminding me

![final wip](https://github.com/comfyanonymous/ComfyUI/assets/1151589/d34b96b9-7722-4ba0-9d03-ca171ad388ae)

**Comment by morphles** [comment]
Guys, I'm not sure I get exactly what you mean, but I think it might be same stuff I'm seeing often, with my hi rest extra custom sampling (though I was seeing it with more regular stuff, I always assumed I'm having cfg to high or maybe unlucky seed). Things like "pink/purple lens flares" or just complete sharp edged pink/purple blotch. If it's vae and not stuff from sampling than it likely lead me astray with trying to tweak settings when I did not need to... For this toxic forest, I see that similar weird color, though here it seems a bit more bluish. So I just wanted to make sure that it's this that is considered wrong?

**Comment by knigitz** [comment]
This needs higher priority, and a fix.  Currently, loading an image and piping it through a VAE Encode/VAE Decode with any VAE input will produces artifacts in the image output.  This even adds artifacts in the areas that are masked not to change during an inpaint process.

Very simple workflows have already been provided above.  This issue is months old and needs proper scoping.

Here is a jeep before and after comfy vae process:
![image](https://github.com/comfyanonymous/ComfyUI/assets/3522823/1469c1ef-24c6-4f80-88b5-10c4baf472b9)

This inpaints very poorly:
![image](https://github.com/comfyanonymous/ComfyUI/assets/3522823/8d73aa2a-e6e5-44c2-a984-640689a4f1d8)

**Comment by WASasquatch** [comment]
Look how small that text is though. How do you expect the integrity of that to hold up when it's at 64^2 noisy?

> This needs higher priority, and a fix. Currently, loading an image and piping it through a VAE Encode/VAE Decode with any VAE input will produce artifacts in the image output. This even adds artifacts in the areas that are masked not to change during an inpaint process.
> 
> Very simple workflows have already been provided above. This issue is months old and needs proper scoping.
> 
> Here is a jeep before and after comfy vae process: ![image](https://user-images.githubusercontent.com/3522823/259493595-1469c1ef-24c6-4f80-88b5-10c4baf472b9.png)
> 
> This inpaints very poorly: ![image](https://user-images.githubusercontent.com/3522823/259494173-8d73aa2a-e6e5-44c2-a984-640689a4f1d8.png)

**Comment by Lalimec** [comment]
I thought this was an expected behavior in inpainting pipeline, no matter what there is always a step of noise in the original image. a1111 combines original image with the inpainted one to keep the details and such same.

**Comment by alenknight** [comment]
is this still a thing?  was hoping would be fixed... there's quite a lot of uses where vaedecode/encode is critical and shouldn't be breaking images.

**Comment by Piezoid** [comment]
VAE encode and decode is an inherently lossy compression process. It's more or less accurate depending on the VAE training process.

Could it be that A1111's img2img blends the original image on top of the output in order to compensate for this?

I reproduced the VAE encode/decode using the HF diffusers library:

<details>

```python
import torch
from diffusers import AutoencoderKL
from diffusers.image_processor import VaeImageProcessor
from diffusers.utils import load_image, pt_to_pil

torch.set_grad_enabled(False)
vae = AutoencoderKL.from_pretrained('madebyollin/sdxl-vae-fp16-fix').to('cuda')
vae_scale_factor = 2 ** (len(vae.config.block_out_channels) - 1)
image_processor = VaeImageProcessor(vae_scale_factor=vae_scale_factor)

img = image_processor.preprocess(load_image('/tmp/239745270-e5965c37-c53b-4e06-9bee-41b816514f68.jpeg')).to('cuda') # The sample image from @aaron9412 above
vae_encode_output = vae.encode(img)
vae_decode_output = vae.decode(vae_encode_output.latent_dist.mean)
pt_to_pil(vae_decode_output.sample.cpu())[0] #Shows up in notebook, otherwise call .save('dest.png')
```
</details>

Left: HF diffusers, right: ComfyUI
<img src="https://github.com/comfyanonymous/ComfyUI/assets/3748582/6f077be2-8a5f-4232-a2f9-15defb06b543" width="294px"> <img src="https://github.com/comfyanonymous/ComfyUI/assets/3748582/fceeaaa9-fc80-49a3-9ae7-e55c98884dfd" width="334px">
They look about the same.

**Comment by WASasquatch** [comment]
> VAE encode and decode is an inherently lossy compression process. It's more or less accurate depending on the VAE training process.
> 
> Could it be that A1111's img2img blends the original image on top of the output in order to compensate for this?
> 
> I reproduced the VAE encode/decode using the HF diffusers library:
> 
> Left: HF diffusers, right: ComfyUI <img alt="" width="294px" src="https://user-images.githubusercontent.com/3748582/284024882-6f077be2-8a5f-4232-a2f9-15defb06b543.png"> <img alt="" width="334px" src="https://user-images.githubusercontent.com/3748582/284025032-fceeaaa9-fc80-49a3-9ae7-e55c98884dfd.png"> They look about the same.

ComfyUI version does strangely look more cartoons despite being more or less the same. Strange effect its giving me.

**Comment by Piezoid** [comment]
You're right, there is some subtle differences. A ~7x boost on the symmetric differences give this:
 <img src="https://github.com/comfyanonymous/ComfyUI/assets/3748582/fc028376-1a24-477b-bd2b-c4281cc96803" width="294px">
[krita file doing the difference](https://github.com/comfyanonymous/ComfyUI/files/13401837/diff.zip)

**Comment by nurhesen** [comment]
Still an issue. Everything is messed up with images smaller than 1000px
